File: src/backend/studio-reservation/src/main/java/org/taonaw/studio_reservation/domain/model/shared/DateTimeRange.java
package org.taonaw.studio_reservation.domain.model.shared;

import lombok.EqualsAndHashCode;
import lombok.Getter;
import lombok.NonNull;
import org.taonaw.studio_reservation.domain.shared.Assertion;

import java.time.LocalDateTime;

@Getter
@EqualsAndHashCode
public class DateTimeRange {
    private final LocalDateTime startDateTime;
    private final LocalDateTime endDateTime;

    public DateTimeRange(LocalDateTime startDateTime, LocalDateTime endDateTime) {
        Assertion.required(startDateTime);
        Assertion.required(endDateTime);

        if (!startDateTime.isBefore(endDateTime)) {
            throw new IllegalArgumentException("終了日時が開始日時以前です。");
        }

        this.startDateTime = startDateTime;
        this.endDateTime = endDateTime;
    }

    public boolean isIn(@NonNull DateTimeRange other) {
        return !startDateTime.isBefore(other.startDateTime) && !endDateTime.isAfter(other.endDateTime);
    }

    public boolean isOverlapped(@NonNull DateTimeRange other) {
        return startDateTime.isBefore(other.endDateTime) && other.startDateTime.isBefore(endDateTime);
    }

    public boolean isPassed(@NonNull LocalDateTime currentDateTime) {
        return !startDateTime.isBefore(currentDateTime);
    }
}

File: src/backend/reservation/src/main/kotlin/org/taonaw/reservation/domain/model/shared/DateTimeRange.kt
package org.taonaw.reservation.domain.model.shared

import java.time.LocalDateTime

interface DateTimeRange {
    val start: LocalDateTime
    val end: LocalDateTime

    fun hasStartSeconds(): Boolean {
        return start.second != 0 || start.nano != 0
    }

    fun hasEndSeconds(): Boolean {
        return end.second != 0 || end.nano != 0
    }

    fun isIn(other: DateTimeRange): Boolean {
        return other.start <= start && end <= other.end
    }

    fun isIn(timeRange: TimeRange): Boolean {
        if (timeRange.isAllDay()) return true
        if (isIn(timeRange.toDateTimeRange(start.toLocalDate()))) return true
        return false
    }

    fun isOverlapping(other: DateTimeRange): Boolean {
        if (start == other.start && end == other.end)
            return true
        return start < other.end && other.start < end
    }

    fun overlappingCount(others: List<DateTimeRange>): Int {
        return others.count { it.isOverlapping(this) }
    }

    fun isPassed(currentDateTime: LocalDateTime): Boolean {
        return start <= currentDateTime
    }

    override fun equals(other: Any?): Boolean
    override fun hashCode(): Int
    override fun toString(): String
}

data class BasicDateTimeRange(
        override val start: LocalDateTime,
        override val end: LocalDateTime) : DateTimeRange {
}